internal/devicescale/impl_desktop.go

```diff
@@ -0,0 +1,45 @@
+// Copyright 2021 The Ebiten Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !android && !ios && !js
+// +build !android,!ios,!js
+
+package devicescale
+
+import (
+	"github.com/hajimehoshi/ebiten/v2/internal/glfw"
+)
+
+func monitorAt(x, y int) *glfw.Monitor {
+	// GLFW only provides monitor x and y coordinates but no reliable sizes.
+	// So the "correct" monitor is the one closest to x, y to the bottom right.
+	// This usually works, but in some exceptional layouts may return the wrong monitor.
+	// Thus, this function is best called with the top-left coordinates of an actual monitor if possible.
+	var best *glfw.Monitor
+	var bestScore int
+	for _, mon := range glfw.GetMonitors() {
+		mx, my := mon.GetPos()
+		if x < mx || y < my {
+			continue
+		}
+		score := (x - mx) + (y - my)
```

Comment:
Can we determine the monitor by

```
vm := m.GetVideoMode()
mx <= x && x < mx + vm.Width && my <= y && y < my + vm.Height
```

Reply:
Specifically this does not work when using a scale > 1 in xrandr (or in Cinnamon, 150% fractional scaling + HiDPI in the dropdown below), as then vm.Width will be _too small_ and if the window origin is near the bottom right corner, it won't match at all.

This is also due to https://github.com/glfw/glfw/issues/1961

Reply:
I see, so probably we should fix `uidriver/glfw/ui.go` as we have the same issue. Is that correct?

https://github.com/hajimehoshi/ebiten/blob/dbe101a8efc7bbd6695201bf07846b2e20fceb38/internal/uidriver/glfw/ui.go#L247-L258

I'd be happy if you leave a comment about this.

Reply:
Added comments.

FYI attached a pic demonstrating the current and new approach's problem:

![bitmap](https://user-images.githubusercontent.com/251568/132899424-8c7581fa-25bb-4cd3-87c7-5b632977e727.png)

- Top: red screen has a 1.5x scale (i.e. video mode is 50% bigger than the window system resolution). You get something like this if you set Cinnamon's scale to 150% but do NOT enable HiDPI. Green screen is normal.
  - The point at the X is in the red screen's "video mode" rectangle but not the window system's. getMonitorFromPosition would falsely map it to the red monitor, not the green one.
- Bottom: all screens have 1:1 scale, a row of two at the top, another monitor bottom of both.
  - The point at the X is on the blue screen, but has a shorter distance to the green screen's top-left corner and thus would be falsely mapped to the green one.

I consider the lower scenario more contrived, as most multi-monitor setups are just a row of screens, so I'd prefer the lower bug if we must have any. So it may make sense to rewrite the getMonitorFromPosition logic the same way as monitorAt (or in fact, just factor devicescale.monitorAt into its own package and use it from both).

Reply:
> The point at the X is in the red screen's "video mode" rectangle but not the window system's. getMonitorFromPosition would falsely map it to the red monitor, not the green one.

So, if video mode's size is divided by the scale 1.5, could we get an exact layout and 'x' was in the green rectangle?

Reply:
friendly ping

Reply:
Oh, hadn't seen the question.

If we divide video mode size by the scale, we could probably get things right - but then we need xrandr to answer monitorAt, which we need to answer xrandr. A messy circular dependency I don't want to get into unless you insist.

At the very least I'd prefer keeping things as is and independent until someone from GLFW responds to my bug.

Reply:
Thanks, so let's not divide the reported sizes by the scale, as GLFW is in a messy state. If GLFW fixes the issue and Ebiten applies that version, let's revisit this issue.

What do you think about this:

> Then, what we have to do here is just check whether the monitor's position exactly matches the arguments or not, right? I'm so sorry it was late to realize this, but this insights should be applied to internal/uidriver/glfw instead.

Reply:
Done that now; indeed, all callers either pass a monitor origin, or the coordinate system origin for "don't care, don't know".